File: GitFix.py

import yaml
import os
import sys
from colorama import Fore
from openai import OpenAI
import random
import time


from github_api import Github_API_Wrapper
from redis import Redis_Wrapper as Redis
from grammar_correction import generate_gramatically_correct_content
def read_config_file():
    config = None
    if not os.path.isfile("config.yaml"):
        print(Fore.RED + 'Error: Config file does not exist!')
        exit(1)
    with open("config.yaml", "r") as cf:
        try:
            config = yaml.safe_load(cf)
        except yaml.YAMLError as exc:
            print(Fore.RED + 'Error in reading config file!')
            print(exc)
            exit(1)
    return config
def gitfix(owner, repo, printer, demo_mode = False):
    printer.print(f"Processing the repository {owner}/{repo}")
    config = read_config_file()
    original_repo = Github_API_Wrapper(owner, repo, config["github-token"])
    original_repo.get_items(printer=printer, demo_mode=demo_mode)
    if not (len(original_repo.items) > 0):
        printer.print("""Gitfix could not discover any files in the repositoy.\n
                      Make sure you inputed your repository name correctly and your repository is indexed in Github search engine.\n
                      If your repository is not indexed, please wait a while until Github indexes your repository.
                      """)    
    printer.print("Forking the repository.")
    try:
        forked_repo = original_repo.fork()
        time.sleep(1)
    except :
        printer.print("Forking process failed, aborting!")
        return
  
    forked_repo.create_a_ref_from_default_branch("gitfix")
    printer.print("Establishing redis connection.")
    try:
        redis = Redis(config["upstash-redis-url"], config["upstash-redis-token"])
        unupdated_items = redis.get_difference(path,original_repo.items)
        original_repo.items = unupdated_items
        forked_repo.items = unupdated_items
    except:
        printer.print("Cannot connect to redis, aborting!")
        return
    openai_client = OpenAI(api_key=config["openai-key"])
    indexes = [random.randint(0, len(original_repo.items)-1) for i in range(config['files-per-run'])]
    printer.print("Selecting files to update.")
    printer.print("Selected files:")
    for k in indexes:
        printer.print(f"\t {original_repo.items[k][0]}")
    for k in indexes:
        printer.print(f"Updating {original_repo.items[k][0]}")
        file_content = original_repo.get_item_content(k)
        corrected_content = generate_gramatically_correct_content(openai_client, file_content)
        forked_repo.update_file_content(k, corrected_content)
        redis.insert(path, forked_repo.items[k])
    printer.print("Creating PR request")
    original_repo.create_PR(forked_repo)
if __name__ == "__main__":

    config = read_config_file()

    path = config["github-repo"]
    path_tokens = path.split("/")
    owner, repo = path_tokens[-2], path_tokens[-1]
    class Local_Printer:
        def print(self ,str):
            print(str)
    printer = Local_Printer()
    gitfix(owner, repo, printer=printer)
     
    

File: src/app/api/gitfix/[owner]/[repo]/gitfix.ts
import GithubAPIWrapper from "./github_api";
import RedisWrapper from './redis_wrapper';
import grammar_correction from './grammar_correction';

async function* gitfix(owner: string, repo: string, demo_mode: boolean, config: any): AsyncGenerator {

  let path = owner + '/' + repo;
  const sleep = async (ms: number) =>
    (new Promise(resolve => setTimeout(resolve, ms)))
  let redis = new RedisWrapper(config["upstash-redis-url"], config["upstash-redis-token"])

  if (demo_mode) {
    let prev_changed_files = await redis.getMembers(owner)
    if (prev_changed_files.length > 3) {
      yield "Error: Gitfix has already corrected 3 files in repositories of this user in demo mode. Please install gitfix for unlimited use. For installation, see https://github.com/upstash/gitfix for more.\n\n"
      return
    }
  }


  //TODO: make the 3 files per user check here 

  yield `Processing the repository ${path}\n\n`;

  let originalRepo = new GithubAPIWrapper(owner, repo, config['github-token'])
  await originalRepo.getItems()

  let logs = "Discovering items:\n\n"
  for (let i = 0; i < originalRepo.items.length; ++i) {
    logs += '- ' + originalRepo.items[i].path + '\n\n';
  }


  if (originalRepo.items.length == 0) {
    yield `Error: Gitfix could not discover any files in the repositoy.
    Make sure you inputed your repository name correctly and your repository is indexed in Github search engine.
    If your repository is not indexed, please wait a while until Github indexes your repository.\n\n`
    return
  } else {
    yield `Gitfix discovered ${originalRepo.items.length} files in the repository.\n\n`
  }

  yield 'Forking the repository. \n\n'
  let forkedRepo: GithubAPIWrapper;
  try {
    forkedRepo = await originalRepo.fork()
    await sleep(500);
  } catch (e) {
    yield "Error: Forking process failed, aborting!\n\n"
    console.log(e)
    return
  }
  try {
    let unupdatedItems = await redis.getDifference(path, originalRepo.items)
    if (unupdatedItems.length < 1) {
      yield "Gitfix has already corrected all grammar errors in the repository.\n\n"
      return
    }
    console.log(unupdatedItems)
    originalRepo.items = unupdatedItems
    forkedRepo.items = unupdatedItems
  } catch (e) {
    yield "Error: Cannot connect to Redis, aborting!\n\n"
    console.log(e)
    return
  }
  //TODO: smarter file selection     

  try {
    forkedRepo.details['default_branch'] = originalRepo.details['default_branch'];
    await forkedRepo.createARefFromDefaultBranch("gitfix")
  } catch (e) {
    console.log(e)
    yield `Error: Gitfix could not create a new branch for changes.\n
    This is most likely due to delays in completing the forking process on Github's end.\n
    Please try again in a minute.\n`;
    return;
  }

  let fileIndexes: Set<number> = new Set();
  if (demo_mode) {
    let promises: Promise<string | void>[] = [];
    let sizes: { [id: number]: number;[size: string]: number }[] = [];
    for (let i = 0; i < originalRepo.items.length; i++) {
      promises.push(
        originalRepo.getItemContent(i).then(function (content) {
          sizes.push({ size: content.length, id: i });
        })
      )
    }
    await Promise.all(promises);
    sizes.sort((a, b) => a.size - b.size)
    console.log(sizes)
    let index = 0;
    for (let i = 0; i < sizes.length; i++) {
      console.log(sizes[i])
      if (sizes[i].size > 100 && sizes[i].size < 15000) {
        fileIndexes.add(sizes[i].id);
      }
      if (fileIndexes.size >= config['files-per-run']) {
        break;
      }
    }
  }
  else {
    while (fileIndexes.size < Math.min(config['files-per-run'], originalRepo.items.length)) {
      fileIndexes.add(Math.floor(Math.random() * (originalRepo.items.length)))
    }
  }
  if (fileIndexes.size > 0) {
    yield "Selecting files to update.\n\n"
  } else {
    yield "Error: Gitfix could not find an unupdated file that is sufficiently long(more than 100 chars). Aborting.\n"
    return
  }
  console.log(logs)
  const indexes = Array.from(fileIndexes);
  yield "Selected files:\n\n"
  for (let i = 0; i < indexes.length; i++) {
    yield `- ${originalRepo.items[indexes[i]].path}\n\n`
  }
  let promises: Promise<void>[] = []
  let yields: string[] = []
  let errored = 0
  for(let i = 0; i < indexes.length; i ++){
    let index =  indexes[i];
    yield `Processing ${originalRepo.items[index].path}\n\n`;
    let file_content = await originalRepo.getItemContent(index);
    console.log(`Processing ${originalRepo.items[index].path} size : ${file_content.length}\n\n`)
    if(file_content.length>50){
      console.log(`create promise for ${originalRepo.items[index].path}`)
      promises.push(
        grammar_correction(file_content, config).then(async function(corrected_content){
          console.log(`Updating ${originalRepo.items[index].path} size : ${file_content.length}`)
          yields.push(`Updating ${originalRepo.items[index].path}\n\n`)
          await forkedRepo.updateFileContent(index, corrected_content)
          await redis.insert(owner, forkedRepo.items[index])
          console.log(`resolving ${originalRepo.items[index].path} size : ${file_content.length}`)
        }
        ).catch((e) =>{
          console.log(e)
          yields.push(`Error: File ${originalRepo.items[index].path} is not updated due to limitations in OpenAI API.\nPlease try again in a few minutes.\n`)
          errored += 1;
        })
      ) 
    }
  }
  let ended = false
  Promise.all(promises).then(() => {ended = true});
  let increments = 0;
  while(!ended){
    console.log(`waiting for all promises to finish: elapsed time: ${increments}`)
    await sleep(1000);
    if (yields.length == 0){
      yield '\u200B';
    }
    else{
      yield yields.shift()
    }
    increments += 1
  }
  if (errored == indexes.length) {
    //yield an error here
  }
  const targetRepo = originalRepo;
  const prKey = targetRepo.owner + targetRepo.repo + "PR";
  type PR = {
    link: string;
    number: number;
}
  let previousPRlist: PR[] = await redis.getMembers(prKey);
  if (previousPRlist.length > 0) {
    console.log(`previously existing pr`)
    let previousPR = previousPRlist[0]
    console.log(previousPR)
    if(!(await targetRepo.checkIfMerged(previousPR.number))){
      yield `Success: Gitfix found a previous PR that includes suggested grammar corrections and appended changes to it. You can see the created request [here](${previousPR.link}).`
      return
   }
  }

  yield "Creating PR request.\n";
  let response = await targetRepo.createPR(forkedRepo);
  let content = await response.json()
  console.log(content)
  if(response.ok){

    redis.insert(prKey, ({"link" : content.html_url, "number": content.number}))
    yield `Success: Gitfix created a PR that includes suggested grammar corrections. You can see the created request [here](${content.html_url}).`
  }else{
    yield `Error: Something went wrong during PR creation, please try again in a minute.`
  }


}

export default gitfix